### PR TITLE
Refactor work loop

### PR DESCRIPTION
**Pull request by fw42**
https://github.com/Shopify/resque/pull/17

@dylanahsmith already reviewed this. @Sirupsen, mind taking a look too?

Will try it in production for a while first before merging.

### CODE REVIEW

**Comment by fw42** [comment]
don't know why CI is failing, something about JSON. I'm pretty sure it has nothing to do with this PR.

File: lib/resque/worker.rb [hunk 1]
```diff
@@ -269,11 +234,29 @@ def work(interval = 5.0, &block)
...
       unregister_worker
     rescue Exception => exception
-      unless exception.class == SystemExit && !@child && run_at_exit_hooks
-        log_with_severity :error, "Failed to start worker : #{exception.inspect}"
+      return if exception.class == SystemExit && !@child && run_at_exit_hooks
+      log_with_severity :error, "Failed to start worker : #{exception.inspect}"
+      unregister_worker(exception)
+    end
+
+    def work_one_job(job = nil, &block)
```
Comment (Sirupsen) [inline lib/resque/worker.rb]: Do you need a test for this public interface that it's now taking an argument for a job?

**Comment by Sirupsen** [comment]
Other than testing the new public method, LGTM

**Comment by grosser** [comment]
@fw42 can you release this ? ... I want to build a new plugin on this ...

**Comment by fw42** [comment]
Hmmm not sure if I have permission to release new gem versions. I have never released a resque gem before. @steveklabnik, any process around that that I need to care about?

**Pull request merged**